[HLSL] Implement HLSL intialization list support

llvm-beanz opened the pull request:
This PR implements HLSL's initialization list behvaior as specified in the draft language specifcation under
[*Decl.Init.Agg*](https://microsoft.github.io/hlsl-specs/specs/hlsl.html#Decl.Init.Agg).

This behavior is a bit unusual for C/C++ because intermediate braces in initializer lists are ignored and a whole array of additional conversions occur unintuitively to how initializaiton works in C.

The implementaiton in this PR generates a valid C/C++ initialization list AST for the HLSL initializer so that there are no changes required to Clang's CodeGen to support this. This design will also allow us to use Clang's rewrite to convert HLSL initializers to valid C/C++ initializers that are equivalent. It does have the downside that it will generate often redundant accesses during codegen. The IR optimizer is extremely good at eliminating those so this will have no impact on the final executable performance.

There is some opportunity for optimizing the initializer list generation that we could consider in subsequent commits. One notable opportunity would be to identify aggregate objects that occur in the same place in both initializers and do not require converison, those aggregates could be initialized as aggregates rather than fully scalarized.

Closes #56067

llvm-beanz committed:
[HLSL] Implement HLSL intialization list support

This PR implements HLSL's initialization list behvaior as specified in
the draft language specifcation under
[*Decl.Init.Agg*](https://microsoft.github.io/hlsl-specs/specs/hlsl.html
#Decl.Init.Agg).

This behavior is a bit unusual for C/C++ because intermediate braces in
initializer lists are ignored and a whole array of additional
conversions occur unintuitively to how initializaiton works in C.

The implementaiton in this PR generates a valid C/C++ initialization
list AST for the HLSL initializer so that there are no changes required
to Clang's CodeGen to support this. This design will also allow us to
use Clang's rewrite to convert HLSL initializers to valid C/C++
initializers that are equivalent. It does have the downside that it
will generate often redundant accesses during codegen. The IR optimizer
is extremely good at eliminating those so this will have no impact on
the final executable performance.

There is some opportunity for optimizing the initializer list
generation that we could consider in subsequent commits. One notable
opportunity would be to identify aggregate objects that occur in the
same place in both initializers and do not require converison, those
aggregates could be initialized as aggregates rather than fully
scalarized.

Closes #56067

llvm-beanz commented:
A couple notes for myself:
* I need to write some AST tests
* Also need to write tests to cover bit fields and unions.

inbelic commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -2576,3 +2576,162 @@ void SemaHLSL::processExplicitBindingsOnDecl(VarDecl *VD) {
+  llvm::SmallVector<Expr *, 16> ArgExprs;
+  bool ExcessInits = false;
+  for (Expr *Arg : Init->inits())
+    BuildIntializerList(SemaRef, Ctx, Arg, ArgExprs, DestTypes, ExcessInits);
```
It seems like this can fail if `S.CreateBuiltinArraySubscriptExpr` or `S.BuildFieldReferenceExpr` returns an invalid expr? Do we care if that is the case? Or if `CastInitializer` fails.

llvm-beanz replied:
If any of those cases fail, they will generate errors, which will (eventually) stop processing. We can continue through all the arguments though and let the error be surfaced after the full initializer expression is processed.

llvm-beanz commented:
note to self: https://hlsl.godbolt.org/z/1Tq59r89W

hekota reviewed (commented):
Few typos and questions, otherwise looks good!
Typos in title and description: intialization, implementaiton

llvm-beanz committed:
Update clang/test/SemaHLSL/Language/InitLists.hlsl

Doh!

Co-authored-by: Finn Plummer <50529406+inbelic@users.noreply.github.com>

llvm-beanz committed:
Update clang/lib/Sema/SemaHLSL.cpp

I swear I can spell...

Co-authored-by: Helena Kotas <hekotas@microsoft.com>

llvm-beanz commented on clang/test/SemaHLSL/Language/InitLists.hlsl:
```diff
@@ -0,0 +1,69 @@
+  TwoFloats F2 = {1}; // expected-error{{too few initializers in list for type 'TwoFloats' (expected 2 but found 1)}}
+  TwoFloats F3 = {1,2,3}; // expected-error{{too many initializers in list for type 'TwoFloats' (expected 2 but found 2)}}
+
+  int2 Something = {1.xxx}; // expected-error{{too many initializers in list for type 'int2' (aka 'vector<int, 2>') (expected 2 but found 0)}}
```
You're correct, this should be 3 as well.

llvm-beanz committed:
Update clang/test/SemaHLSL/Language/InitLists.hlsl

Co-authored-by: Finn Plummer <50529406+inbelic@users.noreply.github.com>

llvm-beanz committed:
Updates to fix diagnostics and adjust to new upstream changes

More updates coming to handle additional PR review.

llvm-beanz committed:
Update to support classes with base classes

Also added tests for bitfield members to verify correct code generation
for initializing bitfield members or initializing new objects from
bitfields.

../clang/test/CodeGenHLSL/BasicFeatures/InitLists.hlsl

llvm-beanz commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -2576,3 +2576,162 @@ void SemaHLSL::processExplicitBindingsOnDecl(VarDecl *VD) {
+  }
+
+  auto It = ArgExprs.begin();
+  // GenerateInitLists will always return an InitListExpr here, because the
+  // scalar case is handled above.
```
The very top of the function handles that case by checking if `Init->getType()->isScalarType()` and early returning.

llvm-beanz committed:
Update clang/test/CodeGenHLSL/BasicFeatures/InitLists.hlsl

Co-authored-by: Helena Kotas <hekotas@microsoft.com>

llvm-beanz committed:
Fix formatting

farzonl commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -11678,9 +11678,12 @@ static void AnalyzeImplicitConversions(
-  // precision loss, because C++11 narrowing already handles it.
-  bool IsListInit = Item.IsListInit ||
-                    (isa<InitListExpr>(OrigE) && S.getLangOpts().CPlusPlus);
+  // precision loss, because C++11 narrowing already handles it. HLSL's
```
The HLSL comment is completely different from the c++ narrowing comment. Feel like it should be on its own line  so as not to communicate that they are related.

llvm-beanz committed:
Split HLSL's comment from C++'s

llvm-beanz commented:
I did find a bug in this for handling data structures with resources... I'm working on it and will update the PR when I have a fix.

farzonl commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -3010,3 +3010,183 @@ void SemaHLSL::processExplicitBindingsOnDecl(VarDecl *VD) {
+
+  if (DestTypes.size() != ArgExprs.size() || ExcessInits) {
+    int TooManyOrFew = ExcessInits ? 1 : 0;
+    SemaRef.Diag(Init->getBeginLoc(), diag::err_hlsl_incorrect_num_initializers)
```
The `note_ovl_candidate_bad_list_argument` note does almost everything you want except error you are returning a false here anyways, do we need a special diagnostic for this? If so does it need to be hlsl specifc. seems like it could be usefull in other cases  maybe replace `err_excess_initializers` with one that can `say too few` or `too many`?

llvm-beanz replied:
In basically all C-based languages "too few" initializers isn't an error, they're just assumed to default to 0 or default initialization. `note_ovl_candidate_bad_list_argument ` is close, but it is a note, not an error. We could potentially share diagnostic strings, but I'm not sure how useful that is.

The main reason I didn't extend `err_excess_initializers`, is because this is definitely an HLSL feature that is on the chopping block. It's hugely problematic and irregular, and it'll be easier to remove support for it someday in the future if it has a unique diagnostic.

That is not strictly a requirement, but boy do I want to burn this feature with maximum prejudice... if only it were not relied on so much.

hekota reviewed (approved):
LGTM!

llvm-beanz committed:
Fix initialization with resources in structs

This was a bit tricker than I expected, but I think I came up with a
reasonably clever solution.

In HLSL, user-defined data types have aggregate initialization, not
constructors _except_ that some of the builtin types we do model
constructors for. This is actually useful!

In my earlier change I updated DeclCXX so that HLSL non-implicit
classes are always marged as Aggregates. This ends up being not
quite right, because there are some implicit types that should be
aggregates, and others that shouldn't. For example, the implicit
cbuffer-layout types should be aggregates so that we can flatten
them, but resources shouldn't be because we really don't want to
flatten them.

In the update, whether or not an HLSL type is an aggregate is keyed
off having non-implicit "special" members (constructors & operators).
This is more correct.

Aggregate types get flattened out for casting and initialization, while
non-Aggregate types (basically just resources) get left unflattened and
we attempt copy-initialization on them.

Next problem: I was getting some odd conflicting diagnostics when
argument conversion or copy-initialization fails, so I refactored
BuildInitializerList and CastInitializer to return success/failure so
that we can propagate that up and fail if the argument->destination
type fails without then also complaining about the number of
initializers.

tex3d reviewed (commented):
This is interesting.

I wonder though what it looks like in the AST when initializing something with the result of a function call that return an aggregate that must be split up across different boundaries?  Or what about expressions with side-effects?  How does it not duplicate sub-expressions while also not creating temporary variables to capture results?

Shouldn't there be an AST test so we can verify the way it transforms the code?

```hlsl
struct Foo {
    float f1;
    float f2;
};

Foo makeFoo() {
    Foo foo = {1, 2};
    return foo;
}

struct TwoFoo {
    Foo foo1, foo2;
};

TwoFoo testSplitFoo() {
    TwoFoo twoFoo = {0, makeFoo(), 3};
    return twoFoo;
}
```

tex3d commented:
> How does it not duplicate sub-expressions while also not creating temporary variables to capture results?

The way I thought of to comprehensively solve this in a clear general way was to:
- create a new temporary with an anonymous struct type containing the exact sequence of expression types from the initializer
- initialize that with a flat initializer list containing the original expressions
- then refer to that temporary in expressions in a new rewritten initializer list matching the target type layout.

llvm-beanz commented:
> > How does it not duplicate sub-expressions while also not creating temporary variables to capture results?
> 
> The way I thought of to comprehensively solve this in a clear general way was to:
> 
> * create a new temporary with an anonymous struct type containing the exact sequence of expression types from the initializer
> * initialize that with a flat initializer list containing the original expressions
> * then refer to that temporary in expressions in a new rewritten initializer list matching the target type layout.
> 
> Or perhaps that can be wrapped in a helper function instead.

I think the simpler approach is to wrap any expression that can possibly side-effect in an `OpaqueValueExpr`. That's how we handle side-effecting `inout` arguments. It allows us to emit the expression once during codegen to a temporary that we can then refer to multiple times because the OVE will appear in the AST in multiple initializer list arguments.

llvm-beanz committed:
Add OVE to capture side effecting arguments in init list

llvm-beanz committed:
Handle side-effecting argument expressions

This slightly tweaks the AST formulation to wrap any potentially
side-effecting initialization list expression in an OpaqueValueExpr so
that during codegen we can ensure it only gets emitted once.

I've made a slightly hacky but minimal change in CodeGen to then emit
OpaqueValueExprs in InitLists during aggregate initialization emission.

I weighed the tradeoff between an AST-level representation or this
CodeGen change. If HLSL were going to retain this initialization list
behavior long term, I'd probably add a new AST node to represent HLSL
initialization lists and restructure how we generate the AST, but I
think that would be a lot more code to maintain, and since the goal is
to remove this quirk from the language I don't think it is the best
solution.

The change as written isolates most of the weirdness of HLSL in
CGHLSLRuntime and is a massively smaller change than a new AST node for
initialization lists. This will also ensure that Clang's AST-based
analysis for initialization lists will continue to be accurate for HLSL
without any additional updates required.

llvm-beanz committed:
Add tests for side-effecting arguments and AST structure.

V-FEXrt commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -3173,9 +3173,18 @@ bool SemaHLSL::TransformInitList(const InitializedEntity &Entity,
+    if (E->HasSideEffects(Ctx)) {
+      QualType Ty = E->getType();
+      if (auto *RTy = Ty->getAs<RecordType>())
+        E = new (Ctx) MaterializeTemporaryExpr(Ty, E, E->isLValue());
```
Unless there is some magic I'm missing, this is memory leak right? 3182 reassigns `E` without freeing? 
Does `OpaqueValueExpr` take ownership of the old `E`?

llvm-beanz replied:
All AST node memory is owned by the ASTContext (hence the placement new-style syntax).

`E` is always owned by the ASTContext, I'm just shifting where in the AST it appears from being the top-level expression in the argument list to being a child of the MaterializeTemporary or OpaqueValue expression, which (at the end of the function) all gets rewritten over the original InitListExpr's Init expressions.

bogner commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -2582,17 +2582,20 @@ static void BuildFlattenedTypeList(QualType BaseTy,
+      if (RD->isStandardLayout())
+        RD = RD->getStandardLayoutBaseWithFields();
+
+      // For types that we shouldn't decompose (unios and non-aggregates), just
```
```suggestion
      // For types that we shouldn't decompose (unions and non-aggregates), just
```

V-FEXrt commented on clang/lib/CodeGen/CGExprAgg.cpp:
```diff
@@ -1730,6 +1731,16 @@ void AggExprEmitter::VisitCXXParenListOrInitListExpr(
+  // emit these we need to emit the opaque values before we emit the argument
+  // expressions themselves. This is a little hacky, but it prevents us needing
+  // to do a bigger AST-level change for a language feature that we need
+  // deprecate in the near future.
```
Can you/should you reference an issue here to link this to the "future deprecation"?

llvm-beanz replied:
I don't think it makes sense to file an issue because even when we depreciate it in the language we'll still need to support it in the compiler for... a long time.

In the sema code I referred to 0005-strict-initializer-lists.md & https://github.com/microsoft/hlsl-specs/pull/325, which are language features that will require changing this quirk of HLSL. I can refer to those two here as well.

bogner commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -3010,3 +3013,193 @@ void SemaHLSL::processExplicitBindingsOnDecl(VarDecl *VD) {
+  if (Ty->isScalarType() || (Ty->isRecordType() && !Ty->isAggregateType()))
+    return CastInitializer(S, Ctx, E, List, DestTypes);
+
+  if (auto *ATy = Ty->getAs<VectorType>()) {
```
I find these abbreviations a bit confusing. `ATy` is a vector type? `VTy` is an array type? Might be best to use longer names here.

bogner commented on clang/lib/AST/DeclCXX.cpp:
```diff
@@ -1462,6 +1462,14 @@ void CXXRecordDecl::addedMember(Decl *D) {
+  // relevant HLSL feature proposals that will depend on this changing:
+  // * 0005-strict-initializer-lists.md
+  // * https://github.com/microsoft/hlsl-specs/pull/325
+  if (getLangOpts().HLSL)
+    data().Aggregate = data().UserDeclaredSpecialMembers == 0;
```
Does this comment match what this is doing? If I'm reading this correctly we're using the fact that there are no user declared constructors/destructors/assignment operators to decide that this is an aggregate - I guess that's because users aren't allowed to write those in HLSL so we can use it as a proxy?

llvm-beanz replied:
Yea... I'll update the comment too. We can't strictly use `implicit` because some of the implicit types do behave like aggregates.

bogner reviewed (approved):
This all looks reasonable to me. A couple of nitpicks and a question inline.

llvm-beanz committed:
Update clang/lib/Sema/SemaHLSL.cpp

Co-authored-by: Justin Bogner <mail@justinbogner.com>

llvm-beanz committed:
Updates based on PR review feedback

llvm-beanz committed:
Update tests based on changes that have merged to main
../clang/test/SemaHLSL/Language/ElementwiseCast-errors.hlsl
